cmd/kubectl-k8ssandra/register/registration.go

```diff
@@ -0,0 +1,143 @@
+package register
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/k8ssandra/k8ssandra-client/pkg/registration"
+	configapi "github.com/k8ssandra/k8ssandra-operator/apis/config/v1beta1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/result"
+)
+
+type RegistrationExecutor struct {
+	DestinationName  string
+	SourceKubeconfig string
+	DestKubeconfig   string
+	SourceContext    string
+	DestContext      string
+	SourceNamespace  string
+	DestNamespace    string
+	ServiceAccount   string
+	Context          context.Context
+}
+
+func getDefaultSecret(saName, saNamespace string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      saName + "-secret",
+			Namespace: saNamespace,
+			Annotations: map[string]string{
+				"kubernetes.io/service-account.name": saName,
+			},
+		},
+		Type: corev1.SecretTypeServiceAccountToken,
+	}
+}
+func getDefaultServiceAccount(saName, saNamespace string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      saName,
+			Namespace: saNamespace,
+		},
+	}
+
+}
+
+func (e *RegistrationExecutor) RegisterCluster() result.ReconcileResult {
```

Comment:
issue: Why are we returning ReconcileResult instead of just error? This return value isn't used with controller-runtime's Reconciler and we should not bring such logic here for no reason.

Reply:
I know that. The plan was basically to distinguish between errors (fatal) and retries, because this is interfacing with an async system and we need to be able to retry while providing updates to the user on what's happening.

How do you want to do this?

Reply:
I can create some new structure which is like a reconcile.Result, but it'll basically have the same semantics, so I figured re-use would make life easier.

Reply:
In this case, there's no need to go so complex. Your usage of reconcile.Result is basically equivalent to (completed bool, err error). If the bool is true, we're done. If not and err == nil, we keep going. If err != nil, there's an error. 

```go
completed, err := executor.RegisterCluster()
if err != nil {
  return err // Or log an error and return, just normal error handling
}

if completed {
  break
}
```

Or do any other !completed -> continue, otherwise return etc. Or include the completed in the for loop statement, but the principle is the same.

You can use ``PollUntilContextTimeout`` and others if you wish to have a timeout instead of count of retries. That's how the start/stop commands work for example.

Reply:
I think we also need to be able to signal that a non-recoverable error has occurred too and there's no point continuing to retry. Right now, we could do it the way you suggest. and treat errors as retryable errors, and panics as non-retryable ones.

But you also want me to remove the panics too. Which will make recoverable and non-recoverable errors indistinguishable. 

So which one do we want to do here - 
1. panic instead of returning errors when the error is non-recoverable, and get rid of the use of reconcile.Result?
2. return result.Requeue for recoverable errors and result.Error for non-recoverable errors, and get rid of the panics?
3. Create some new form of recoverable/non-recoverable error handling mechanism specifically for this application?

Reply:
> Create some new form of recoverable/non-recoverable error handling mechanism specifically for this application?

@burmanm and I have discussed and agreed that we will do this one.